phi_four.py

```diff
@@ -0,0 +1,237 @@
+    """example of training loop of model"""
+    # create your optimizer and a scheduler
+    optimizer = optim.Adadelta(model.parameters(), lr=1)
+    scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=500)
+    # let's use tqdm to see progress
+    pbar = tqdm(range(epochs), desc=f"loss: N/A")
+    for i in pbar:
+        # gen simple states
+        z = torch.randn((N_BATCH, N_UNITS))
+        phi = model.inverse_map(z)
+        target = target_distribution_s(phi, m_sq, l, shift)
+        output = model(phi)
+
+        model.zero_grad() # get rid of stored gradients
+        loss = shifted_kl(output, target)
+        loss.backward() # calc gradients
+
+        optimizer.step()
+        scheduler.step(loss)
+
+        if (i%50) == 0:
+            pbar.set_description(f"loss: {loss.item()}")
+
+def sample(model, a_left, b_right, n_large, target_length):
+    r"""
+    Sample using Metroplis-Hastings algorithm from a large number of phi configurations.
+    We calculate an A = min(1, \frac{\tilde p(phi^i)}{p(phi^i)} \frac{p(phi^j)}{\tilde p(phi^j)})
+    Where i is the index of the current phi, and j is the index of the update proposal phi.
+    We then generate a random number u, and if u <= A then we accept the update and i = j for the
+    next proposal, phi^i is added to the chain of samples, and a new j is picked. If the update
+    is rejected, then i=i for the next proposal and a new j is picked. We continue until the chain
+    has the desired length.
+    """
+
+    with torch.no_grad(): # don't want gradients being tracked in sampling stage
+        z = torch.randn((n_large, N_UNITS)) # random z configurations
+        phi = model.inverse_map(z) # map using trained model to phi
+        # p_tilde = torch.exp(model.forward(phi)) # probabilities of generated phis using trained model
+        log_ptilde = model.forward(phi)
+        # p = torch.exp(-target_distribution_s(phi, m_sq, l, shift)) # probabilities of phis from target pdf
+        S = -target_distribution_s(phi, m_sq, l, shift)
+        chain_len = 0 # intialise current chain length
+        sample_distribution = torch.Tensor(target_length, N_UNITS) # intialise tensor to store samples
+        accepted, rejected = 0,0 # track accept/reject statistics
+        i = np.random.randint(n_large) # random index to start sampling from configurations tensor
+        used = [] # track which configurations have been added to the chain
+        used.append(i) 
+        while chain_len < target_length:
+            j = np.random.randint(n_large) # random initial phi^j for update proposal
+            while j in used: # make sure we don't pick a phi we have already used
+                j = np.random.randint(n_large)
+            # ratio = (p_tilde[i]/p_tilde[j]) * (p[j]/p[i]) # calculate the ratio in A
+            exponent = log_ptilde[i] + S[j] - log_ptilde[j] - S[i]
+            P_accept = np.exp(float(exponent)) # much faster if you tell it to use a float
```

Comment:
I would imagine this is because you have a torch tensor (exponent) which numpy has to work the typing out for. Possibly, but maybe not `exponent.exp().item()` would be faster or the same as what you're doing

although they are similar remember numpy != pytorch and so using numpy functions on pytorch objects means that the object is probably copied along the way.. (more memory more time etc.)

Reply:
nope yours is faster! 👍 
```python
In [1]: import torch                                                            

In [2]: import numpy as np                                                      

In [3]: a = torch.rand(1)                                                       

In [4]: a                                                                       
Out[4]: tensor([0.2411])

In [5]: %timeit min(1, a.exp().item())                                          
2.6 µs ± 21 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit min(1, np.exp(float(a)))                                        
2.09 µs ± 18.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit min(1, float(a.exp().item()))                                   
2.72 µs ± 36.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
mind you...
```python
In [8]: from math import exp                                                    

In [9]: %timeit min(1, exp(float(a)))                                           
596 ns ± 3.19 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
this all said, we should worry about optimising later